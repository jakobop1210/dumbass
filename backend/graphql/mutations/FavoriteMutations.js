const { GraphQLNonNull, GraphQLString, GraphQLInt } = require('graphql')
const { UserType } = require('../types/UserType')
const User = require('../../models/User')

const addFavorite = {
    type: UserType,
    args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        type: { type: new GraphQLNonNull(GraphQLString) },
        targetId: { type: new GraphQLNonNull(GraphQLInt) }
    },
    resolve: async (parent, args) => {
        try {
            const user = await User.findOne({ username: args.username })
            if (!user) {
                throw new Error('User not found.')
            }
            // Check if the favorite already exists for the user
            const favoriteExist = user.favorites.some((favorite) => {
                return (
                    favorite.type === args.type &&
                    favorite.targetId === args.targetId
                )
            })
            if (favoriteExist) {
                throw new Error('Favorite already exists.')
            }
            // If the favorite doesn't exist, add it
            const favoriteObj = {
                type: args.type,
                targetId: args.targetId
            }
            user.favorites.push(favoriteObj)
            // Save the updated user
            return user.save()
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

const removeFavorite = {
    type: UserType,
    args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        type: { type: new GraphQLNonNull(GraphQLString) },
        targetId: { type: new GraphQLNonNull(GraphQLInt) }
    },
    resolve: async (parent, args) => {
        try {
            const user = await User.findOne({ username: args.username })
            if (!user) {
                throw new Error('User not found.')
            }
            // Find the index of the favorite to remove
            const favoriteIndex = user.favorites.findIndex((favorite) => {
                return (
                    favorite.type === args.type &&
                    favorite.targetId === args.targetId
                )
            })

            if (favoriteIndex === -1) {
                throw new Error('Favorite not found.')
            }
            // Remove the favorite from the user's favorites array
            user.favorites.splice(favoriteIndex, 1)
            return await user.save()
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

module.exports = { addFavorite, removeFavorite }
