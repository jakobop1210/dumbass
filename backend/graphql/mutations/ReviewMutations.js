const {
    GraphQLNonNull,
    GraphQLString,
    GraphQLInt,
    GraphQLBoolean
} = require('graphql')
const { ReviewType } = require('../types/ReviewType')
const User = require('../../models/User')
const Review = require('../../models/Review')
const Artist = require('../../models/Artist')
const Song = require('../../models/Song')

const addReview = {
    type: ReviewType,
    args: {
        userName: { type: new GraphQLNonNull(GraphQLString) },
        content: { type: GraphQLString },
        rating: { type: new GraphQLNonNull(GraphQLInt) },
        targetType: { type: new GraphQLNonNull(GraphQLString) },
        targetId: { type: new GraphQLNonNull(GraphQLInt) }
    },
    resolve: async (parent, args) => {
        try {
            if (args.targetType !== 'artist' && args.targetType !== 'song') {
                throw new Error(
                    'Invalid targetType. It should be "artist" or "song".'
                )
            }

            // Check if user exists
            const user = await User.findOne({ username: args.userName })
            if (!user) {
                throw new Error('User not found.')
            }

            // Check if the user already has a review for this target
            const existingReview = await Review.findOne({
                userName: args.userName,
                targetType: args.targetType,
                targetId: args.targetId
            })

            if (existingReview) {
                throw new Error('User already has a review for this target.')
            }

            // Check if rating is between 0 and 5
            if (args.rating < 0 || args.rating > 5) {
                throw new Error('Rating should be between 0 and 5.')
            }

            // Check if targetId exists
            let target
            if (args.targetType === 'artist') {
                target = await Artist.findOne({ id: args.targetId })
            } else if (args.targetType === 'song') {
                target = await Song.findOne({ id: args.targetId })
            }

            if (!target) {
                throw new Error(`Target with id ${args.targetId} not found.`)
            }

            // Create a new review object
            const review = new Review({
                userName: args.userName,
                content: args.content,
                rating: args.rating,
                targetType: args.targetType,
                targetId: args.targetId
            })

            // Save the review to the database
            const savedReview = await review.save()

            // Update the average rating and number of ratings for the target
            target.average_rating =
                (target.average_rating * target.number_of_ratings +
                    args.rating) /
                (target.number_of_ratings + 1)
            target.number_of_ratings += 1
            await target.save()

            return savedReview
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

const deleteReview = {
    type: GraphQLBoolean,
    args: {
        userName: { type: new GraphQLNonNull(GraphQLString) },
        targetType: { type: new GraphQLNonNull(GraphQLString) },
        targetId: { type: new GraphQLNonNull(GraphQLInt) }
    },
    resolve: async (parent, args) => {
        try {
            if (!['artist', 'song'].includes(args.targetType)) {
                throw new Error(
                    'Invalid targetType. It should be "artist" or "song".'
                )
            }

            const review = await Review.findOneAndDelete({
                userName: args.userName,
                targetType: args.targetType,
                targetId: args.targetId
            })

            if (!review) {
                throw new Error('Review not found.')
            }

            const target =
                args.targetType === 'artist'
                    ? await Artist.findOne({ id: args.targetId })
                    : await Song.findOne({ id: args.targetId })

            if (!target) {
                throw new Error(`Target with id ${args.targetId} not found.`)
            }

            target.average_rating =
                (target.average_rating * target.number_of_ratings -
                    review.rating) /
                Math.max(target.number_of_ratings - 1, 1)

            target.number_of_ratings = Math.max(target.number_of_ratings - 1, 0)

            await target.save()

            return true
        } catch (error) {
            throw new Error(error.message)
        }
    }
}

module.exports = { addReview, deleteReview }
