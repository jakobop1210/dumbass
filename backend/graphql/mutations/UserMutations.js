const { UserType } = require('../types/UserType')
const User = require('../../models/User')
const bcrypt = require('bcrypt')
const { GraphQLNonNull, GraphQLString } = require('graphql')

/**
 * GraphQL Mutation Field - addUser.
 *
 * Creates a new user with a username and password. This mutation ensures that the username is unique 
 * and hashes the password before saving it to the database.
 *
 * @type {Object} GraphQL mutation configuration object.
 * @property {UserType} type - The GraphQL type that this mutation will return.
 * @property {Object} args - Arguments required for this mutation.
 * @property {function} resolve - The resolver function to execute the mutation.
 */
const addUser = {
    type: UserType,
    args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        password: { type: new GraphQLNonNull(GraphQLString) }
    },
    resolve: async (parent, args) => {
        const existingUsername = await User.findOne({
            username: args.username
        })
        if (existingUsername) {
            throw new Error('Username already taken.')
        }
        const saltRounds = 10
        const hashedPassword = await bcrypt.hash(args.password, saltRounds)

        const user = new User({
            username: args.username,
            password: hashedPassword,
            favorites: []
        })

        return user.save()
    }
}

/**
 * GraphQL Mutation Field - loginUser.
 *
 * Validates a user's login credentials. This mutation checks if the username exists and 
 * if the provided password matches the hashed password stored in the database.
 *
 * @type {Object} GraphQL mutation configuration object.
 * @property {UserType} type - The GraphQL type that this mutation will return.
 * @property {Object} args - Arguments required for this mutation.
 * @property {function} resolve - The resolver function to execute the mutation.
 */
const loginUser = {
    type: UserType,
    args: {
        username: { type: new GraphQLNonNull(GraphQLString) },
        password: { type: new GraphQLNonNull(GraphQLString) }
    },
    resolve: async (parent, args) => {
        // Check if username exists
        const user = await User.findOne({ username: args.username })
        if (!user) {
            throw new Error('User not found.')
        }
        // Check if password is correct
        const isPasswordValid = await bcrypt.compare(
            args.password,
            user.password
        )
        if (!isPasswordValid) {
            throw new Error('Incorrect password.')
        }
        return user
    }
}

module.exports = { addUser, loginUser }
