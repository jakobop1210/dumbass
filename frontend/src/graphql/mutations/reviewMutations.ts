import { gql } from '@apollo/client'

/**
 * GraphQL mutation: ADD_REVIEW.
 *
 * This mutation is used to add a new review to a song or artist in the database.
 * It takes the username, review content, rating, target type (song or artist), and target ID as parameters.
 *
 * The mutation returns the details of the newly added review, including the username, review content, rating, target type, and target ID.
 *
 * @param {string} $userName - The name of the user adding the review.
 * @param {string} [$content] - The content of the review.
 * @param {number} $rating - The rating given by the user.
 * @param {string} $targetType - The type of the target being reviewed (either 'song' or 'artist').
 * @param {number} $targetId - The unique identifier of the target (song or artist).
 * @returns {Object} The review object containing userName, content, rating, targetType, and targetId.
 */
export const ADD_REVIEW = gql`
    mutation (
        $userName: String!
        $content: String
        $rating: Int!
        $targetType: String!
        $targetId: Int!
    ) {
        addReview(
            userName: $userName
            content: $content
            rating: $rating
            targetType: $targetType
            targetId: $targetId
        ) {
            userName
            content
            rating
            targetType
            targetId
        }
    }
`
