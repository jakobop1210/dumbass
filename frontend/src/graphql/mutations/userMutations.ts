import { gql } from '@apollo/client'

export const ADD_USER = gql`
    mutation addUser($username: String!, $email: String!, $password: String!) {
        addUser(username: $username, email: $email, password: $password) {
            username
            email
        }
    }
`

export const LOGIN_USER = gql`
    mutation loginUser($username: String!, $password: String!) {
        loginUser(username: $username, password: $password) {
            username
        }
    }
`
