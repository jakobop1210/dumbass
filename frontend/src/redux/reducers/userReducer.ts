import { createReducer } from '@reduxjs/toolkit';
import { setUserLogin, setUserLogout, setUserName } from '../actions/userActions';

interface UserState {
    loggedIn: boolean;
    username: string | null;
}

const initialState: UserState = {
    loggedIn: false,
    username: null,
};

const userReducer = createReducer(initialState, (builder) => {
    builder
        .addCase(setUserLogin, (state) => {
            state.loggedIn = true;
        })
        .addCase(setUserLogout, (state) => {
            state.loggedIn = false;
            state.username = null; // Nullify username
        })
        .addCase(setUserName, (state, action) => {
            state.username = action.payload; // Set username
        });
});

export default userReducer;
