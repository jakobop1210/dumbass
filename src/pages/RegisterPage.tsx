import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import InputField from '../components/molecules/InputField'
import Button from '../components/molecules/Button'
import { customToast } from '../lib/utils'

/**
 * RegisterPage component - A user interface for account registration
 *
 * This component provides form fields for user registration and performs
 * simple front-end validation (password matching). Upon successful validation,
 * it redirects users to the login page and provides a success toast notification.
 * 
 * @returns {JSX.Element}
 */
export default function RegisterPage(): JSX.Element {
    const navigate = useNavigate()
    const [username, setUsername] = useState('')
    const [email, setEmail] = useState('')
    const [password, setPassword] = useState('')
    const [confirmPassword, setConfirmPassword] = useState('')

    /**
     * registerAccount - Event handler for form submission
     *
     * This function is triggered upon form submission. It performs a simple
     * front-end validation check to ensure passwords match and provides
     * relevant toast notifications. Upon successful validation, it redirects
     * the user to the login page and provides a success toast notification.
     * 
     * @param {React.FormEvent} e - Form event
     */
    function registerAccount(e: React.FormEvent) {
        e.preventDefault()
        if (password !== confirmPassword) {
            customToast('error', 'The passwords does not match')
            return
        }
        navigate('/login')
        customToast('success', 'User successfully created')
    }

    return (
        <div className='w-screen flex justify-center'>
            <div className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <div className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <h2 className='text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Create an account
                    </h2>
                </div>

                <div className='mt-6 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={registerAccount}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='email'
                            type='email'
                            title='Email'
                            value={email}
                            onChange={setEmail}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <InputField
                            id='confirmPassword'
                            type='password'
                            title='Confirm Password'
                            value={confirmPassword}
                            onChange={setConfirmPassword}
                            required
                        />
                        <Button title='Create Account' type='submit' />
                    </form>

                    <p className='mt-10 text-center text-sm text-blueGray'>
                        Already have an account?
                        <a
                            onClick={() => navigate('/login')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Login here
                        </a>
                    </p>
                </div>
            </div>
        </div>
    )
}
