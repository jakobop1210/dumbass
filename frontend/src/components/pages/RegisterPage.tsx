import { useNavigate } from 'react-router-dom'
import { useState } from 'react'
import InputField from '../atoms/InputField'
import Button from '../atoms/Button'
import { customToast } from '../../lib/utils'
import { ADD_USER } from '../../graphql/mutations/userMutations'
import { useMutation } from '@apollo/client'

/**
 * `RegisterPage` Component.
 *
 * This component presents a registration form for creating a new user account. 
 * It includes input fields for username, password, and password confirmation, along with a submit button.
 * The form performs validation to ensure that the entered passwords match.
 *
 * Upon form submission, a GraphQL mutation is used to attempt creating a new user. 
 * Success or failure feedback is provided to the user through custom toast messages.
 * On successful account creation, the user is navigated to the login page.
 * In case of errors such as a username conflict, an appropriate error message is displayed.
 *
 * The page also provides a link to the login page for users who already have an account.
 *
 * @returns {JSX.Element} The rendered registration page with a form for creating a new user account.
 */
export default function RegisterPage(): JSX.Element {
    const navigate = useNavigate()
    const [username, setUsername] = useState('')
    const [password, setPassword] = useState('')
    const [confirmPassword, setConfirmPassword] = useState('')
    const [addUser] = useMutation(ADD_USER)

    /**
     * Handles the registration form submission.
     * Validates the form data and uses the `ADD_USER` mutation for creating a new user.
     *
     * @param {React.FormEvent} e - The form event.
     */
    async function registerAccount(e: React.FormEvent) {
        e.preventDefault()
        if (password !== confirmPassword) {
            customToast('error', 'The passwords does not match')
            return
        }
        try {
            const { data } = await addUser({
                variables: { username, password }
            })

            if (data.addUser) {
                customToast('success', 'User successfully created')
                navigate('/login')
            } else {
                customToast('error', 'Failed to create the user')
            }
        } catch (error) {
            customToast('error', 'Username is already taken')
        }
    }

    return (
        <main className='w-screen flex justify-center'>
            <section className='flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white'>
                <header className='sm:mx-auto sm:w-full sm:max-w-sm'>
                    <h2 className='text-center text-3xl font-bold leading-9 tracking-tight text-blueGray'>
                        Create an account
                    </h2>
                </header>

                <div className='mt-6 sm:mx-auto sm:w-full sm:max-w-sm'>
                    <form className='space-y-6' onSubmit={registerAccount}>
                        <InputField
                            id='username'
                            type='username'
                            title='Username'
                            value={username}
                            onChange={setUsername}
                            required
                        />
                        <InputField
                            id='password'
                            type='password'
                            title='Password'
                            value={password}
                            onChange={setPassword}
                            required
                        />
                        <InputField
                            id='confirmPassword'
                            type='password'
                            title='Confirm Password'
                            value={confirmPassword}
                            onChange={setConfirmPassword}
                            required
                        />
                        <Button title='Create Account' type='submit' />
                    </form>

                    <footer className='mt-10 text-center text-sm text-blueGray'>
                        Already have an account?
                        <button
                            onClick={() => navigate('/login')}
                            className='ml-2 font-semibold leading-6 text-green hover:cursor-pointer'>
                            Login here
                        </button>
                    </footer>
                </div>
            </section>
        </main>
    )
}
