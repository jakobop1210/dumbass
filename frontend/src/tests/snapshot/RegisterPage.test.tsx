import { render } from '@testing-library/react'
import { BrowserRouter as Router } from 'react-router-dom'
import { MockedProvider } from '@apollo/client/testing'
import RegisterPage from '../../components/pages/RegisterPage'
import { Provider } from 'react-redux';
import { ADD_USER } from '../../graphql/mutations/userMutations'
import { vi } from 'vitest'
import userReducer from '../../redux/reducers/userReducer';
import { configureStore } from '@reduxjs/toolkit';

vi.mock('react-router-dom', () => ({
    useNavigate: () => vi.fn(),
    // eslint-disable-next-line
    BrowserRouter: require('react-router-dom').BrowserRouter
}))

const mockStore = configureStore({
    reducer: {
        user: userReducer,
    },
    preloadedState: {
        user: { loggedIn: true, username: 'testUser' },
    },
});

const mocks = [
    {
        request: {
            query: ADD_USER,
            variables: { username: 'testuser', password: 'password123' }
        },
        result: {
            data: {
                addUser: {
                    username: 'testuser'
                }
            }
        }
    }
]

test('RegisterPage snapshot', () => {
    const { asFragment } = render(
        <Provider store={mockStore}>
            <MockedProvider mocks={mocks} addTypename={false}>
                <Router>
                    <RegisterPage />
                </Router>
            </MockedProvider>
        </Provider>
    )
    expect(asFragment()).toMatchSnapshot()
})
