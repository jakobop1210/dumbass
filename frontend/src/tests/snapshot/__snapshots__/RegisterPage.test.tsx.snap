// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RegisterPage snapshot 1`] = `
<DocumentFragment>
  <main
    class="w-screen flex justify-center"
  >
    <section
      class="flex h-4/5 sm:w-[32rem] w-4/5 flex-col justify-center px-8 py-10 my-10 rounded-lg md:px-20 bg-white"
    >
      <header
        class="sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <h2
          class="text-center text-3xl font-bold leading-9 tracking-tight text-blueGray"
        >
          Create an account
        </h2>
      </header>
      <div
        class="mt-6 sm:mx-auto sm:w-full sm:max-w-sm"
      >
        <form
          class="space-y-6"
        >
          <main>
            <label
              class="block text-sm font-medium leading-6 text-blueGray"
              for="username"
            >
              Username
            </label>
            <input
              class="block w-full h-12 rounded-md border-0 py-1.5 px-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red sm:text-sm sm:leading-6 mt-2"
              id="username"
              name="username"
              placeholder="Username"
              required=""
              type="username"
              value=""
            />
          </main>
          <main>
            <label
              class="block text-sm font-medium leading-6 text-blueGray"
              for="password"
            >
              Password
            </label>
            <input
              class="block w-full h-12 rounded-md border-0 py-1.5 px-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red sm:text-sm sm:leading-6 mt-2"
              id="password"
              name="password"
              placeholder="Password"
              required=""
              type="password"
              value=""
            />
          </main>
          <main>
            <label
              class="block text-sm font-medium leading-6 text-blueGray"
              for="confirmPassword"
            >
              Confirm Password
            </label>
            <input
              class="block w-full h-12 rounded-md border-0 py-1.5 px-2 text-blueGray shadow-sm ring-1 ring-inset ring-gray-300 placeholder:text-gray-300 focus:ring-2 focus:ring-inset focus:ring-red sm:text-sm sm:leading-6 mt-2"
              id="confirmPassword"
              name="confirm password"
              placeholder="Confirm Password"
              required=""
              type="password"
              value=""
            />
          </main>
          <button
            class="flex w-full justify-center rounded-md px-3 py-1.5 text-sm font-semibold leading-8 text-white shadow-[0_4px_9px_-4px_#8FC0A9] undefined"
            type="submit"
          >
            Create Account
          </button>
        </form>
        <footer
          class="mt-10 text-center text-sm text-blueGray"
        >
          Already have an account?
          <button
            class="ml-2 font-semibold leading-6 text-green hover:cursor-pointer"
          >
            Login here
          </button>
        </footer>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
